### PR TITLE
Tracking area solution

### LINKED ISSUE #17: Improve mouse tracking
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/CircularProgress/issues/17)

We currently do mouse tracking as a square around the circle, but that leads to the cancel state being activated even when the mouse is not over the circle. We should only activate the cancel state when the mouse is actually over the circle.

https://stackoverflow.com/questions/38963011/nstrackingarea-for-custom-shapes

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)
### Submitted pull Requests
- [#22 Tracking area solution](https://issuehunt.io/r/sindresorhus/CircularProgress/pull/22)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/CircularProgress/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by eonist**
Fixes https://github.com/sindresorhus/CircularProgress/issues/17

Proposal for adding TrackingArea that works with CGPath. Ive used this solution in various projects and  I feel its almost optimal. Gets away with not having a tracking area variable but adds a bool flag, which I have yet to be able to remove. More detail about why the different methods and variables exist and how they work in tandem in the code-comments.

Feel free to make requests: 

- Which Linter style to use, Vanilla swift-lint?
- I know you like less comments so, tone that down maybe?
- Abstract it into The CustomButton instead maybe?
- Clean up the project a bit? It's a bit long in the tooth, IMO some +Extensions would do it good
- There seems to be some margin inside .bounds, I will add that to the path variable
- Other?

I will donate the profits to a charitable case. :)
- Would be great to wrap this up before 8th of April. 😅 But however long it takes of course.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#17: Improve mouse tracking](https://issuehunt.io/repos/153606623/issues/17)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by sindresorhus** [comment]
Great! Yes, I think the comments are a bit too much. Try to limit them to only when you need to explain the “why”.

**Comment by sindresorhus** [comment]
I think “Trackable” is a bit too generic. Maybe “MouseTrackable” or something?

**Comment by eonist** [comment]
Did the changes:

- Remove none-essential comments
- Update File name
- Add more robust hitTest point code (supports over/down)

mousetrackable_gif_1

A bit unsure what the extra code in `mouseExited` actually is for, but It can be re-inserted easily:

```swift
progressLabel.isHidden = isIndeterminate && progress == 0
cancelButton.isHidden = true
```

Also maybe not now, but IMO the project could use a cleanup. Move things into manageable pieces so that it's easier to maintain. Let me know if you want me to take a stab at that at a later point. ✌️

**Comment by boyvanamstel** [comment]
> A bit unsure what the extra code in `mouseExited` actually is for, but It can be re-inserted easily:
> 
> ```swift
> progressLabel.isHidden = isIndeterminate && progress == 0
> cancelButton.isHidden = true
> ```

There are quite a few 'states' the button can be in. The `progressLabel` and `cancelButton` need to be shown and hidden accordingly.

> Also maybe not now, but IMO the project could use a cleanup. Move things into manageable pieces so that it's easier to maintain. Let me know if you want me to take a stab at that at a later point. ✌️

Continuing from above: I agree. The 'utils' and `CustomButton` could be better integrated and a single state maintainer would make it a lot easier to see _why_ stuff gets shown or hidden.

If you could open a separate PR that starts on that, I'm all for it. What do you think @sindresorhus?

**Comment by eonist** [comment]
@boyvanamstel I’ll integrate the states as they were before and commit. Then make another pr for a suggestion on how to clean up.

**Comment by eonist** [comment]
@boyvanamstel ✅ added the changes

**Comment by sindresorhus** [comment]
> Which Linter style to use, Vanilla swift-lint?

Yes, SwiftLint. Can you run `swiftlint autocorrect` and then fix the remaining warnings shown in Xcode?

File: Sources/CircularProgress/CircularProgress.swift [hunk 1]
```diff
@@ -35,7 +36,7 @@ public final class CircularProgress: NSView {
 		$0.lineWidth = CGFloat(lineWidth)
 	}
 
-	private lazy var cancelButton = with(CustomButton.circularButton(title: "╳", radius: Double(radius), center: bounds.center)) {
+	private lazy var cancelButton: CustomButton = with(.circularButton(title: "╳", radius: Double(radius), center: bounds.center)) {
```
Comment (sindresorhus) [inline Sources/CircularProgress/CircularProgress.swift]: Don't do unrelated changes.

File: Sources/CircularProgress/protocol/MouseTrackable.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#if os(macOS)
+import Cocoa
+/**
+Makes it possible to utilize mouseOver/mouseOut functionality on a view (Supports custom bezier path)
+*/
+protocol MouseTrackable:class{
```
Comment (sindresorhus) [inline Sources/CircularProgress/protocol/MouseTrackable.swift]: I don't really see the point in making this so loose when it can only be used with `NSView` anyway.

I would make it:

```suggestion
protocol MouseTrackable: NSView {
```

File: Sources/CircularProgress/protocol/MouseTrackable.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	var path:CGPath? {get}
+}
+extension MouseTrackable where Self:NSView{
+	/**
+	* Creates a new tracking area
+	*/
+	internal func createTrackingArea(_ options:NSTrackingArea.Options = [.activeInActiveApp, .mouseEnteredAndExited, .mouseMoved]){
```
Comment (sindresorhus) [inline Sources/CircularProgress/protocol/MouseTrackable.swift]: `internal` is the default. You don't need to specify that.

File: Sources/CircularProgress/protocol/MouseTrackable.swift [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#if os(macOS)
```
Comment (sindresorhus) [inline Sources/CircularProgress/protocol/MouseTrackable.swift]: This is moot. This package is only for macOS.

File: Sources/CircularProgress/protocol/MouseTrackable.swift [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import Cocoa
+/**
+Makes it possible to utilize mouseOver/mouseOut functionality on a view (Supports custom bezier path)
+*/
+protocol MouseTrackable:class{
+	var path:CGPath? {get}
+}
+extension MouseTrackable where Self:NSView{
+	/**
+	* Creates a new tracking area
```
Comment (sindresorhus) [inline Sources/CircularProgress/protocol/MouseTrackable.swift]: It also removes existing tracking areas. That should be mentioned too.

File: Sources/CircularProgress/CircularProgress.swift [hunk 6]
```diff
@@ -415,3 +371,60 @@ public final class CircularProgress: NSView {
...
 }
+
+extension CircularProgress:MouseTrackable{
+	public override func hitTest(_ point: NSPoint) -> NSView? {
+		guard let mousePos = self.window?.mouseLocationOutsideOfEventStream else {return nil}
+		let localMousePos = self.convert(mousePos,from:nil)
+		if let path = path {
+			let contains:Bool = path.contains(localMousePos)
+			return contains ? self : nil
+		}
```
Comment (sindresorhus) [inline Sources/CircularProgress/CircularProgress.swift:382]: The logic for hit testing should reside in the `MouseTrackable` extension. It should handle all the hit logic and here we should only need to supply a `path` to be hit tested against.

File: Sources/CircularProgress/CircularProgress.swift [hunk 7]
```diff
@@ -415,3 +371,60 @@ public final class CircularProgress: NSView {
...
+		if hasMouseEntered == false && viewUnderMouse === self {
+			onOver()
+			hasMouseEntered = true
+		}
+		else if hasMouseEntered && viewUnderMouse !== self{
+			onOut()
+			hasMouseEntered = false
+		}
+	}
+	@objc func onOver(){
```
Comment (sindresorhus) [inline Sources/CircularProgress/CircularProgress.swift]: Why are these using `@objc`?

File: Sources/CircularProgress/protocol/MouseTrackable.swift [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+		trackingAreas.forEach{ trackingArea in
+			self.removeTrackingArea(trackingArea)
+		}
+		let trackingArea:NSTrackingArea = .init(rect: self.bounds, options: options, owner: self, userInfo: nil)
+		self.addTrackingArea(trackingArea)
+	}
+	/**
+	Stores/generates the path to hitTest against
+	*/
+	var path:CGPath? {return nil}
```
Comment (sindresorhus) [inline Sources/CircularProgress/protocol/MouseTrackable.swift]: The naming here is too generic and could easily conflict. Should be `hitTestPath` or something better.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Would be nice if the `MouseTrackable` protocol had custom mouseEntered/mouseExited hooks that would only report when the mouse is actually entered/exited the given `.path`, and it handled the calculations internally. That would make our own custom logic much simpler as we wouldn't have to mix it with the `MouseTrackable` hit testing logic.

File: CircularProgress.xcodeproj/project.pbxproj [hunk 9]
```diff
@@ -85,6 +87,14 @@
 			path = Example;
 			sourceTree = "<group>";
 		};
+		F1FDA5872254CE7D00797872 /* protocol */ = {
+			isa = PBXGroup;
+			children = (
+				F1FDA5882254CE8D00797872 /* MouseTrackable.swift */,
+			);
+			path = protocol;
```
Comment (sindresorhus) [inline CircularProgress.xcodeproj/project.pbxproj:95]: I prefer to not nest it into directories.

File: Sources/CircularProgress/CircularProgress.swift [hunk 10]
```diff
@@ -9,6 +9,7 @@ public final class CircularProgress: NSView {
 	private var finishedObserver: NSKeyValueObservation?
 	private var cancelledObserver: NSKeyValueObservation?
 	private var indeterminateObserver: NSKeyValueObservation?
+	internal var hasMouseEntered:Bool = false/*Required for onOver / onOut to work*/
```
Comment (sindresorhus) [inline Sources/CircularProgress/CircularProgress.swift]: This should be `private`.

File: Sources/CircularProgress/CircularProgress.swift [hunk 11]
```diff
@@ -415,3 +371,60 @@ public final class CircularProgress: NSView {
...
+			return contains ? self : nil
+		}
+		return super.hitTest(point)
+	}
+	var path: CGPath? { return CGPath.init(ellipseIn: bounds.insetBy(dx: 20, dy: 20), transform: nil) }
+	override public func updateTrackingAreas() {
+		guard isCancellable else {
+			return
+		}
+		createTrackingArea([.activeAlways, .mouseMoved, .mouseEnteredAndExited])
```
Comment (sindresorhus) [inline Sources/CircularProgress/CircularProgress.swift]: Why `.activeAlways`?

**Comment by sindresorhus** [comment]
Clicking the cancel button does nothing now, unlike how it works in `master`.

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 8 above)
Comment (sindresorhus) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: Why limit it to a single path. Seems like if we want to make it generally useful, it should be an array of paths?

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 2 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: - it's a protocol not a class, so it can't extend anything else than other protocols. NSView is not a protocol.
- `SomeProtocol:class` in swift just means, it doesn't apply to structs.

File: Sources/CircularProgress/CircularProgress.swift (hunk 1 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: Scouts honour: Leave Things better than you found them 👔 ...but alright.

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 3 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: I can make the extension internal, which would make all encapsulated methods also internal? I could be wrong but internal for me means, only accessible within the .framework.

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 4 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: Yepp. Ill remove it. "Copy-pasta artefact"

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 5 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: Will do

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 3 above)
Comment (sindresorhus) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: > All entities in your code (with a few specific exceptions, as described later in this chapter) **have a default access level of internal if you don’t specify an explicit access level yourself**. As a result, in many cases you don’t need to specify an explicit access level in your code. - https://docs.swift.org/swift-book/LanguageGuide/AccessControl.html

---

> I can make the extension internal, which would make all encapsulated methods also internal?

Nah. I prefer putting access levels on methods themselves, not the whole extension.

File: Sources/CircularProgress/CircularProgress.swift (hunk 6 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift:382]: - The whole "why" this is here was removed when I simplified the comments. 
- Basically the way this is setup is a bit intricate. And it can't really be simplified anymore, I've tried a lot in the past. Like, you have to override hitTest in the class, because hit testing applies to the hierarchy. So if you have a covering NSView it needs to sort of run down the hierarchy. It's complicated. And I'm not explaining it well. 🤷
My comments when I use this solution more generically on other projects. Like for roundButtons
```swift
/**
    * This is the last NSView so we dont forward the hitTest to further descendants, however we could forward the hit test one more step to the CALayer
    * - TODO: ⚠️️ the logic inside this method should be in the Shape, and this method should just forward to the shape, if there is a shape
    */
```

File: Sources/CircularProgress/CircularProgress.swift (hunk 7 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: So you can override methods in extensions. Putting things in extensions avoids MegaControllers. And makes code easier to comprehend. IMO. https://www.natashatherobot.com/using-swift-extensions/  TLDR: While this might not be “traditional”, the super simple use of extensions in Swift can make code super nice and readable.

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 8 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: Im just solving for the CircularProgress project here. I also have this "tech" in another OSS project https://github.com/eonist/Hybrid/blob/master/src/button/button/Button%2BTrackable.swift , were it can gain more things like this. Maybe even become its own Component or maybe a general tutorial on how to do over/Out on custom paths. Which should be so simple 🤷, but isn't 🙁

Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: Ill fix the name. Let me know if you want array support on paths 👍

File: CircularProgress.xcodeproj/project.pbxproj (hunk 9 above)
Comment (eonist) [reply CircularProgress.xcodeproj/project.pbxproj:95]: ok. Flat hierarchy it is.

Comment (eonist) [reply CircularProgress.xcodeproj/project.pbxproj:95]: And a comment on MouseTrackable being more independent. Yeah. Id love that too, but no way around overriding hitTesting inside the class it applies to etc. I have tried many times in the past. This latest stab for Hybrid and CircularProgress I feel is a nice compromise. I also got rid of the storing of NSTracking area in the class, now we just derived them from Apples API

File: Sources/CircularProgress/CircularProgress.swift (hunk 10 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: I think its internal because an extension needs access to it. extensions can't access private's vars.

File: Sources/CircularProgress/CircularProgress.swift (hunk 11 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: I guess it can be replaced with NSTrackingActiveInActiveApp. Or maybe removed all together? I'll have a look.

**Comment by eonist** [comment]
Clicking the cancel button, should be simple enough to fix. Top of my mind its probably due to the new hitTesting regime. Ill have a look 👍

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 3 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: ✅

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 4 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: ✅

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 5 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: ✅

File: Sources/CircularProgress/CircularProgress.swift (hunk 6 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift:382]: I feel this is best: Put the logic inside this method should be in the Shape, and this method should just forward to the shape, if there is a shape. 

But I feel the second best place is to put that code in the NSView where it is now. 

I feel, the code is messy, but other stuff can be cleaned up before this to make things clean. I can do that too, in another PR. If you want. 

Its up to you. If you want it in the protocol I will oblige. ✌️

File: Sources/CircularProgress/CircularProgress.swift (hunk 7 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: Leaving this be.

File: Sources/CircularProgress/protocol/MouseTrackable.swift (hunk 8 above)
Comment (eonist) [reply Sources/CircularProgress/protocol/MouseTrackable.swift]: Fixed the name. ✅ Not doing the multiple path support. Unless you are sure you want it. Let me know. It's  pretty simple.

File: CircularProgress.xcodeproj/project.pbxproj (hunk 9 above)
Comment (eonist) [reply CircularProgress.xcodeproj/project.pbxproj:95]: Skipping this. As I think this is as deep as the rabbit hole goes on this. I could be wrong though. But maybe leave for 🔮 improvement? I feel a refactoring and cleaning up of code would be more appropriate before this tweak 🤷

File: Sources/CircularProgress/CircularProgress.swift (hunk 10 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: ✅ Although it will become internal again when we refactor into using more extensions :D

File: Sources/CircularProgress/CircularProgress.swift (hunk 11 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: ✅ changed to .activeInApp instead.

**Comment by eonist** [comment]
✅ Moved logic to mouseDown (Quick solution, I'm not sure how to make NSButton interactions work, and do we need a button here? this component is pretty custom and NSButton doesn't play well with custom code, in general, could spend hours trying to nail this 🤷)

File: Sources/CircularProgress/CircularProgress.swift (hunk 1 above)
Comment (eonist) [reply Sources/CircularProgress/CircularProgress.swift]: ✅

**Comment by eonist** [comment]
All Should be good on this PR, I followed up on what I interpret as important. If there are things you want me to tweak, let me know and I'll Make an attempt. If your still in thick PR-review traffic. No worries. The summer is long :)

**Comment by sindresorhus** [comment]
I'm sorry for leaving this unattended for so long. I really wanted to simplify the implementation and make it less coupled and more general, but I never had the chance to finish my implementation: https://github.com/sindresorhus/CircularProgress/commit/24a0b5a522ab741d7f6841b1df33283ef3ebe766

I've come to the conclusion that I don't think this complexity is worth the tiny amount of improved usability. Also, this is all moot with SwiftUI which supports shape hit testing built-in.

**Comment by eonist** [comment]
@sindresorhus No worries man, totally agree 👍
